src/core/core/task.cpp

```diff
@@ -182,7 +183,7 @@ void task::exec_internal()
             else
             {
                 _state.compare_exchange_strong(READY_STATE, TASK_STATE_CANCELLED);
-                _spec->on_task_end.execute(this);
+                _spec->on_task_cancelled.execute(this);
```

Comment:
I think it should be: the thread which successfully change state from READY to CANCELLED will execute on_task_cancelled().
And, is on_task_end() should always be called, no matter the task is generally finished or cancelled?